In [ ]:
import pandas as pd
import numpy as np
import pickle # do i need this if I only use the pandas method?
import skfuzzy as fuzz

import ipympl
#%config InlineBackend.figure_format='retina'

import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib_inline
import matplotlib.cm as cm
import seaborn as sns

import warnings 
warnings.filterwarnings("ignore")

Since the t-SNE values provided visible clusters that could not be found via HDBSCAN clustering, I will use t-SNE values to perform fuzzy c-means clustering 

## Fuzzy classification

Multilabel classification is the ability to predict multiple discrete labels/classes per data point input. It should not be confused with multiclass classification, which simply means there are more than two potential classes to predict membership of, but usually only one per data point. Every label is assigned a 0 or 1 for every data point. Although there is no membership probability given, these can be used to rank belongingness. These methods have the advantage of being offered by `sk-learn`.

However, in this case "fuzzy" classification may be more accurate (and more interesting if there are overlapping Roud numbers), as it explicitly outputs degrees of membership. Fuzzy classification is not part of `sk-learn`. I considered the following options:
- `sk-fuzzy` offers it as part of their [fuzzy clustering](https://pythonhosted.org/scikit-fuzzy/api/skfuzzy.cluster.html) but not as a pure classification model. If I take the number of clusters `c` to be 3210 (the output of the best HDBSCAN run) I can get around the limitation of the underlying C-Means model that the number of clusters must be specified in advance. However, the other limitations may still be a problem (i.e. it assumes a uniform cluster shape and struggles with higher-dimensional data).

In [ ]:
import skfuzzy as fuzz
df_classify.reset_index()

level_0  index                      key_name                   name  \
0           0      0          A Robin, Jolly Robin    A Robyn Jolly Robyn   
1           1      1          A Robin, Jolly Robin             (No Title)   
2           2      2              A, U, Hinny Bird       A, U, Hinny Bird   
3           3      3  Adieu to Erin (The Emigrant)          Adieu to Erin   
4           4      4          Agincourt Carol, The  The Song of Agincourt   
...       ...    ...                           ...                    ...   
9968     9968  10108                           NaN     Zeb Tourney's Girl   
9969     9969  10109                           NaN              Zebra Dun   
9970     9970  10110                           NaN     Zen Gospel Singing   
9971     9971  10111                           NaN                Zuleika   
9972     9972  10112                           NaN          The Zulu King   

     version_in_key   bi_file   dt_file    roud  \
0                 A  Perc1185  HEYROBIN     NaN   
1                 B  Perc1185  HEYROBIN     NaN   
2                 A   StoR160       NaN     235   
3                 A   SWMS255       NaN    2068   
4                 A     MEL51  AGINCRT1  V29347   
...             ...       ...       ...     ...   
9968            NaN      LE18  ZEBTURNY    2249   
9969            NaN      LB16  ZEBRADUN    3237   
9970            NaN       NaN  ZENGOSPE     NaN   
9971            NaN       NaN    ZULIKA     NaN   
9972            NaN       NaN  ZULUKING     NaN   

                                                 lyrics  roud_count  \
0     "[F]rom what appears to be the most ancient of...         NaN   
1     71 'Hey, Robin, jolly Robin, 72    Tell me how...         NaN   
2     A, U, hinny burd; The bonny lass o' Benwell, A...         1.0   
3     Oh, when I breathed a last adieu, To Erin's an...         1.0   
4     Deo gracias anglia, Redde pro victoria, 1 Owre...         3.0   
...                                                 ...         ...   
9968  Down in the Tennessee mountains,\nFar from the...         2.0   
9969  We was camped on the plains at the head of the...         1.0   
9970  I once was a Baptist and on each Sunday morn\n...         NaN   
9971  Zuleika was fair to see,\nA fair Persian maide...         NaN   
9972  Oh the Zulu king with the big nose-ring\nFell ...         NaN   

                                 lyric_embed_instructor  \
0     [-0.032208655, -0.0039244993, -0.02159848, 0.0...   
1     [-0.032028995, 0.020379173, -0.016789645, 0.03...   
2     [-0.025857605, 0.010645705, -0.02403562, 0.050...   
3     [-0.043128256, 0.008317871, -0.040352777, 0.01...   
4     [-2.9962948e-05, -0.009317334, -0.017600924, 0...   
...                                                 ...   
9968  [-0.0389951, -0.008086728, -0.058877233, 0.037...   
9969  [-0.044695836, -0.012055198, -0.042596623, 0.0...   
9970  [-0.043211307, -0.0138436, -0.017630804, 0.003...   
9971  [-0.0115272915, -0.029302498, -0.022729361, 0....   
9972  [-0.043027755, -0.010383786, -0.020057475, 0.0...   

                                  tsne_embedding  tsne_hdb_cluster  
0       [-98.88156127929688, 28.498804092407227]               214  
1        [-99.28376770019531, 28.85739517211914]               214  
2      [-74.82328796386719, -26.113561630249023]              3050  
3     [-18.571348190307617, -37.902496337890625]              2787  
4        [-85.41560363769531, 44.82973861694336]                83  
...                                          ...               ...  
9968     [47.74774932861328, 17.576581954956055]              2258  
9969      [49.38885498046875, 6.010506629943848]              2451  
9970          [68.2158203125, 18.50948715209961]              1222  
9971   [15.239205360412598, -15.814308166503906]              2158  
9972    [45.874881744384766, -30.61400032043457]              1243  

[9973 rows x 13 columns]

The data must be input in the format of a 2D array, size (S, N), where N is the number of "data sets" (i don't know if this means data points - [O'Reilly](https://www.oreilly.com/library/view/mastering-machine-learning/9781788621113/6967d36f-e04e-46d3-8c99-e30e2193d464.xhtml) also says it's the number of features!) and S is the number of features in each vector. 

Note: Initially the data was the wrong way round. Each embedding index (768) was clustered instead of each data point (9973) (wrong axis). Per O'Reilly I tranposed the `data` numpy array and the clustering was successful.

In [ ]:
df_classify = pd.read_pickle('df_classify.p')
ncenters = 3210
data = np.array(df_classify['lyric_embed_instructor'].tolist())
data = data.transpose()
data.shape

(768, 9973)

Fit the model. Returns:	
1. cluster centres: `cntr`: "Data for each center(?) along each feature provided" for each of the requested clusters c. 2d array, size (S, c)
2. matrices: `u`: Final fuzzy c-partitioned matrix. `u0`: Initial guess at u matrix. `d`: Final Euclidian distance matrix. 2d array, (S, N)
3. model assessment information: `jm`: "Objective function history" of model performance at each iteration. 1d array, length P. `p`: Number of iterations run. `fpc`: fuzzy partition coefficient.

In [ ]:
cntr, u, u0, d, jm, p, fpc = fuzz.cluster.cmeans(data, ncenters, 2, error=0.005, maxiter=1000, init=None)

Examine the clusters:

In [ ]:
u.shape

(3210, 9973)

In [ ]:
cntr.shape

(3210, 768)

Only 462 clusters were chosen as a first choice for each data point. The largest cluster was assigned 1184 points.

In [ ]:
cluster_membership = np.argmax(u, axis=0) #TODO: check this
clusters, counts = np.unique(cluster_membership, return_counts=True)
counts.max()

1184

The fuzzy partition coefficient (FPC) is a metric of how cleanly the data is described by the model on a scale from 0 to 1, with 1 being the best. My FPC is close to 0 which suggests a poor clustering.

In [ ]:
fpc

0.00031152649052995895

Let's add the first-choice labels and visualise them:

In [ ]:
df_classify['fuzzy_cluster_0'] = cluster_membership
df_classify['fuzzy_cluster_0_label'] = cluster_membership.astype(str) #string version just for plotting labels
df_classify

In [ ]:
#plot
# make non-cluster -1 grey
color_map = {'-1': 'lightgray'}  # nb key is a string
# colours for other cluster labels
num_clusters = ncenters
other_colors = px.colors.qualitative.Light24[:num_clusters - 1]
color_map.update(dict(zip(map(str, range(num_clusters - 1)), other_colors)))

# # Calculate localized centroids for 'roud' annotations
# def calculate_annotations():
#     unique_roud_values = df_classify['roud'].unique()
#     annotations = []

#     for roud_value in unique_roud_values:
#         roud_clusters = df_classify[df_classify['roud'] == roud_value]['fuzzy_cluster_0_label'].unique()
        
#         for cluster_label in roud_clusters:
#             roud_points = df_classify[(df_classify['roud'] == roud_value) & (df_classify['fuzzy_cluster_0_label'] == cluster_label)]
#             if len(roud_points) > 0:
#                 centroid_x = roud_points['tsne_embedding'].str.get(0).mean()
#                 centroid_y = roud_points['tsne_embedding'].str.get(1).mean()
#                 if len(roud_points) > 1:
#                     annotations.append(dict(x=centroid_x, y=centroid_y,
#                                             xref="x", yref="y",
#                                             text=roud_value,
#                                             showarrow=True,
#                                             arrowhead=2,
#                                             ax=0,
#                                             ay=-25))
#                 else:
#                     annotations.append(dict(x=centroid_x, y=centroid_y,
#                                             xref="x", yref="y",
#                                             text=roud_value,
#                                             showarrow=False,
#                                             ax=0,
#                                             ay=0))
#     return annotations

# annotations = calculate_annotations()

# Create the scatter plot
fig = px.scatter(df_classify.sort_values(by='fuzzy_cluster_0'),
                 x=df_classify['tsne_embedding'].str.get(0),
                 y=df_classify['tsne_embedding'].str.get(1),
                 color='fuzzy_cluster_0_label',
                 color_discrete_map=color_map,
                 hover_name='name',
                 hover_data={'roud': True},
                 labels={'fuzzy_cluster_0_label': 'Cluster (Fuzzy, 0)'},
                 title='Whole dataset Fuzzy clustering (instructor embeddings)'
                )

# # Add buttons to toggle annotations
# button_on = dict(label='Annotations On',
#                  method='relayout',
#                  args=[{'annotations': annotations}])

# button_off = dict(label='Annotations Off',
#                   method='relayout',
#                   args=[{'annotations': []}])

# fig.update_layout(updatemenus=[
#     dict(type='buttons', showactive=True, buttons=[button_on, button_off])
# ])

fig.update_layout(xaxis_title='t-SNE X', yaxis_title='t-SNE Y')
fig.update_xaxes(scaleanchor="y", scaleratio=1)
fig.update_yaxes(dtick=20)
fig.update_xaxes(dtick=20)

fig.update_layout(width=1400, height=1100)

fig.show()


Let's now examine the first, second and third choice clusters and their probabilities:

In [ ]:
#TODO: first extract the fuzzy data and append to df

#get a group of songs
df_classify[df_classify['name'] == 'Alouette'] # reveals cluster 2240
df_classify[df_classify['fuzzy_cluster_0'] == 2240]

index                       key_name  \
236     253  Craven Churn-Supper Song, The   
273     290           Do, Do, Pity My Case   
538     570    Jinny Get Your Hoecake Done   
608     645   Lazy Mary (She Won't Get Up)   
641     679  London Bridge Is Falling Down   
909     973        Quaker's Courtship, The   
1075   1152               Three Blind Mice   
1380   1463                            NaN   
2387   2477                            NaN   
2413   2503                            NaN   
2623   2717                            NaN   
2690   2784                            NaN   
2848   2942                            NaN   
4795   4899                            NaN   
4834   4939                            NaN   
5236   5343                            NaN   
5565   5674                            NaN   
5566   5675                            NaN   
5787   5898                            NaN   
5817   5928                            NaN   
6632   6752                            NaN   
6776   6898                            NaN   
6782   6904                            NaN   
7211   7338                            NaN   
8053   8183                            NaN   
8054   8184                            NaN   
8368   8498                            NaN   
9293   9428                            NaN   

                                           name version_in_key   bi_file  \
236               Craven Churn-Supper Song, The              A   BeCo382   
273                        Do, Do, Pity My Case              A    BAF805   
538                                The Hoe-Cake              A   Fus158C   
608          What Will You Give Me if I Get Up?              B      R396   
641               London Bridge Is Falling Down              A      R578   
909                     Quaker's Courtship, The              A      R362   
1075                           Three Blind Mice              A  FSWB413A   
1380                                   Alouette            NaN       NaN   
2387                            Brother Gorilla            NaN       NaN   
2413                          Buckingham Palace            NaN       NaN   
2623                    Character of a Mistress            NaN       NaN   
2690                      Christopher and Alice            NaN       NaN   
2848                        Comin' Thro the Rye            NaN  FSWB140B   
4795                              Hunt the Wren            NaN      K078   
4834                                 I Can Sail            NaN       NaN   
5236                         Jamie, Come Try Me            NaN       NaN   
5565                            Kissin's No Sin            NaN       NaN   
5566      Kissing My Kate (Merry Hae I Been...)            NaN       NaN   
5787                    Let Simon's Beard Alone            NaN       NaN   
5817              Liewer Heindrich (Dear Henry)            NaN  DTholebc   
6632                               Naughty Baby            NaN       NaN   
6776                                Nuts in May            NaN      R561   
6782  O Polly, You Might Have Toy'd and Kissed.            NaN       NaN   
7211                              Perfect Woman            NaN       NaN   
8053                     The Seven Joys of Mary            NaN     FO211   
8054                 The Seven Joys of Mary (2)            NaN     FO211   
8368                               Souling Song            NaN   BGMG408   
9293                      Who Killed Cock Robin            NaN     SKE74   

       dt_file   roud                                             lyrics  \
236        NaN  13471  Be not moved at my strain, For nothing study s...   
273        NaN  11590  Do, do pity my case, In some lady's garden, My...   
538        NaN  16825  Jinny, get your hoecake done, my love, Jinny, ...   
608        NaN   6561  "What will you give me if I get up, If I get u...   
641        NaN    502  London bridge is falling down, Falling down, f...   
909       

In [ ]:
data = pandas.read_pickle(lyrics_dataset.p)

# Perform Fuzzy C-Means clustering
cntr, u, u0, d, jm, p, fpc = fuzz.cluster.cmeans(
    data.T, c=3, m=2, error=0.005, maxiter=1000, init=None
)

# Retrieve the cluster centers
cluster_centers = cntr.T

# Retrieve the membership values
membership_values = u.T